src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/Helpers/ElementModelExtensions.cs

```diff
@@ -29,14 +29,56 @@ internal static class ElementModelExtensions
         // Converts a DataType to a named ElementNode
         internal static ElementNode ToElementNode(this DataType data, string name)
         {
-            var node = ElementNode.FromElement(TypedSerialization.ToTypedElement(data));
+            var node = ElementNode.FromElement(data.ToTypedElement());
             node.Name = name;
             return node;
         }
 
-        // Finds a child ElementNode at a given path
-        internal static ElementNode Find(this ElementNode element, string pathExpression) =>
-            ((ScopedNode)element.Select(pathExpression).First()).Current as ElementNode;
+        // Given the results of a Path Select command, get the first element node
+        internal static ElementNode GetFirstElementNode(this IEnumerable<ITypedElement> nodeList) =>
+            nodeList.First().ToScopedNode().Current.ToElementNode();
+
+        // Used to check return from "select" operation
+        internal static IEnumerable<ITypedElement> CheckNoElements(this IEnumerable<ITypedElement> nodeList)
+        {
+            if (!nodeList.Any())
+            {
+                throw new InvalidOperationException($"No content found");
+            }
+
+            return nodeList;
+        }
+
+        // Used to check return from "select" operation
+        internal static IEnumerable<ITypedElement> CheckMultipleElements(this IEnumerable<ITypedElement> nodeList)
+        {
+            if (nodeList.Count() > 1)
+            {
+                var firstResultLocation = nodeList.First().Location;
+                var expectedLocation = firstResultLocation.Remove(firstResultLocation.LastIndexOf('.'));
+
+                // Multiple results are only allowed on collection types
+                if (nodeList.Any(l =>
+                    !l.Definition.IsCollection ||
+                    !l.Location.StartsWith(expectedLocation, StringComparison.InvariantCultureIgnoreCase)))
+                {
+                    throw new InvalidOperationException($"Multiple matches found");
```

Comment:
I am not sure how the error message matches the check.  We fail if the definition is not a collection type, or if the location does not match the expected location.  But, we report an error of "Mulitple matches found"?

Reply:
The goal of this method is to provide a "builder like" pattern for the "Operation*.cs" classes which define the logic for each FHIRPath Patch operation. The "idea" was to make these classes more clear about what they are checking for instead of hiding everything in the extensions class. Example:

```
Target = ResourceElement
    .Select(Operation.Path)
    .CheckNoElements()
    .CheckMultipleElements()
    .GetFirstElementNode();
```

I'm feeling these could go in their own class called something like "ElementModelTestExtensions" or something similar. The whole `ElementModelExtensions` class has a bit of code smell (sorry) and could be broken up. I was attempting to minimize the changes in the PR, but happy to split it up if you think it improves the quality of the code base. 😓